packages/labs/analyzer/src/lib/javascript/mixins.ts

```diff
@@ -0,0 +1,201 @@
+/**
+ * @license
+ * Copyright 2022 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+/**
+ * @fileoverview
+ *
+ * Utilities for working with mixins
+ */
+
+import ts from 'typescript';
+import {AnalyzerInterface, MixinDeclaration} from '../model.js';
+import {getClassDeclaration} from './classes.js';
+import {createDiagnostic} from '../errors.js';
+import {DiagnosticCode} from '../diagnostic-code.js';
+import {parseNodeJSDocInfo} from './jsdoc.js';
+import {getFunctionLikeInfo} from './functions.js';
+
+const nodeHasMixinHint = (node: ts.Node) =>
+  ts.getJSDocTags(node).some((tag) => tag.tagName.text === 'mixin');
+
+const addDiagnosticIfMixin = (
+  node: ts.Node,
+  hasMixinHint: boolean,
+  message: string,
+  analyzer: AnalyzerInterface
+) => {
+  if (hasMixinHint) {
+    analyzer.addDiagnostic(
+      createDiagnostic({
+        typescript: analyzer.typescript,
+        node,
+        message,
+        code: DiagnosticCode.UNSUPPORTED,
+        category: analyzer.typescript.DiagnosticCategory.Warning,
+      })
+    );
+  }
+  return undefined;
+};
+
+/**
+ * If the given variable declaration was a mixin function, returns a
+ * MixinDeclaration, otherwise returns undefined.
+ *
+ * The mixin logic requires a few important syntactic heuristics to be met in
+ * order to be detected as a mixin.
+ *
+ * If the function is unannotated and does not match the above mixin shape, it
+ * will silently just be analyzed as a simple function and not a mixin. However,
+ * the `@mixin` annotation can be added to produce specific diagnostic errors
+ * when a condition for being analyzed as a mixin is not met.
+ */
+export const maybeGetMixinFromFunctionLike = (
+  fn: ts.FunctionLikeDeclaration,
+  name: string,
+  analyzer: AnalyzerInterface,
+  hasMixinHint = nodeHasMixinHint(fn)
+) => {
+  if (!fn.parameters || fn.parameters.length < 1) {
+    addDiagnosticIfMixin(
+      fn,
+      hasMixinHint,
+      `Expected mixin to have a superClass parameter.`,
+      analyzer
+    );
+    return undefined;
+  }
+  const possibleSuperClasses = fn.parameters.map((p) =>
+    ts.isIdentifier(p.name) ? p.name.text : ''
+  );
+  const functionBody = fn.body;
+  if (functionBody === undefined) {
+    addDiagnosticIfMixin(
+      fn,
+      hasMixinHint,
+      `Expected mixin to have a block function body.`,
+      analyzer
+    );
+    return undefined;
+  }
+  if (!ts.isBlock(functionBody)) {
+    addDiagnosticIfMixin(
+      fn,
+      hasMixinHint,
+      `Expected mixin to have a block function body; arrow-function class ` +
+        `expression syntax is not supported.`,
```

Comment:
FYI, I believe this restriction is just due to the fact that you can't put decorators on class expressions in TS without it complaining (ala [this note](https://lit.dev/docs/composition/mixins/#applying-decorators-in-mixins)); maybe add a TODO here, since this might be overly restrictive for JS users...?